c/xudt_rce.c

```diff
@@ -0,0 +1,526 @@
+
+  struct mol2_cursor_t bytes = input.t->unwrap(&input);
+  // convert Bytes to XudtWitnessInputType
+  XudtWitnessInputType witness_input = make_XudtWitnessInput(&bytes);
+  ScriptVecOptType script_vec =
+      witness_input.t->raw_extension_data(&witness_input);
+
+  uint32_t read_len =
+      mol2_read_at(&script_vec.cur, g_raw_extension_data, RAW_EXTENSION_SIZE);
+  CHECK2(read_len >= script_vec.cur.size, ERROR_INVALID_MOL_FORMAT);
+
+  *var_data = g_raw_extension_data;
+  *var_len = read_len;
+
+  err = 0;
+exit:
+  return err;
+}
+
+// *var_data will point to "Raw Extension Data", which can be in args or witness
+// *var_data will refer to a memory location of g_script or g_raw_extension_data
+int parse_args(int* owner_mode, XUDTFlags* flags, uint8_t** var_data,
+               uint32_t* var_len, uint8_t* hashes, uint32_t* hashes_count) {
+  int err = 0;
+
+  uint64_t len = SCRIPT_SIZE;
+  int ret = ckb_checked_load_script(g_script, &len, 0);
+  CHECK(ret);
+  CHECK2(len <= SCRIPT_SIZE, ERROR_SCRIPT_TOO_LONG);
+
+  mol_seg_t script_seg;
+  script_seg.ptr = g_script;
+  script_seg.size = len;
+
+  mol_errno mol_err = MolReader_Script_verify(&script_seg, false);
+  CHECK2(mol_err == MOL_OK, ERROR_ENCODING);
+
+  mol_seg_t args_seg = MolReader_Script_get_args(&script_seg);
+  mol_seg_t args_bytes_seg = MolReader_Bytes_raw_bytes(&args_seg);
+  CHECK2(args_bytes_seg.size >= BLAKE2B_BLOCK_SIZE, ERROR_ARGUMENTS_LEN);
+
+  *hashes_count = 0;
+  // With owner lock script extracted, we will look through each input in the
+  // current transaction to see if any unlocked cell uses owner lock.
+  *owner_mode = 0;
+  size_t i = 0;
+  while (1) {
+    uint8_t buffer[BLAKE2B_BLOCK_SIZE];
+    uint64_t len2 = BLAKE2B_BLOCK_SIZE;
+    // There are 2 points worth mentioning here:
+    //
+    // * First, we are using the checked version of CKB syscalls, the checked
+    // versions will return an error if our provided buffer is not enough to
+    // hold all returned data. This can help us ensure that we are processing
+    // enough data here.
+    // * Second, `CKB_CELL_FIELD_LOCK_HASH` is used here to directly load the
+    // lock script hash, so we don't have to manually calculate the hash again
+    // here.
+    ret = ckb_checked_load_cell_by_field(buffer, &len2, 0, i, CKB_SOURCE_INPUT,
+                                         CKB_CELL_FIELD_LOCK_HASH);
+    if (ret == CKB_INDEX_OUT_OF_BOUND) {
+      break;
+    }
+    CHECK(ret);
+    CHECK2(*hashes_count < MAX_LOCK_SCRIPT_HASH_COUNT, ERROR_TOO_MANY_LOCK);
+
+    memcpy(&hashes[(*hashes_count) * BLAKE2B_BLOCK_SIZE], buffer,
+           BLAKE2B_BLOCK_SIZE);
+    *hashes_count += 1;
+
+    if (args_bytes_seg.size >= BLAKE2B_BLOCK_SIZE &&
+        memcmp(buffer, args_bytes_seg.ptr, BLAKE2B_BLOCK_SIZE) == 0) {
+      *owner_mode = 1;
+      break;
+    }
+    i += 1;
+  }
+
+  // parse xUDT args
+  if (args_bytes_seg.size < (FLAGS_SIZE + BLAKE2B_BLOCK_SIZE)) {
```

Comment:
Better strictly check `args_bytes_seg.size`? 

So that follow sizes are now allowed: `BLAKE2B_BLOCK_SIZE + {1,2,3}`

Reply:
It's undefined. Currently it's treated as "plain": Simple UDT.